Babyeatingpsychopath refactor

babyeatingpsychopath opened the pull request:
You asked about delayMilliseconds();  This is what I do.  It should keep a somewhat (arguably) more constant time between polls.  In case something takes a weird length of time, you just don't sleep as long.

I didn't change most of the const int XXXX = YYY stuff to #define XXXX YYY because there's a lot and i'm lazy, but that's a thing that lets the compiler really drill down on typos.

babyeatingpsychopath committed:
Refactor for my idiom

Changed data types, to match callers (e.g. gamepad.axis wants int16_t)

babyeatingpsychopath committed:
Make function prototypes match their functions

i changed the functions printAnalog() and readAnalog() without changing their function prototypes.  I don't think the arduino C compiler cares if function prototypes exist, but let's do it the right way.

CoryParsnipson reviewed (commented):
Just FYI, the function prototypes are there because I put the definitions on the bottom of the file and I wanted to call them above. Without them, the compilation fails.

CoryParsnipson commented on code/controller/controller.ino:
```diff
@@ -62,80 +64,71 @@ void setup() {
 
 void loop() {
-  bool is_pressed = false;
+  // bool is_pressed = false;  // unused variable?
```
Yeah used to be used for print messages. You can delete lines and references to this variable.

CoryParsnipson commented on code/controller/controller.ino:
```diff
@@ -62,80 +64,71 @@ void setup() {
+  static unsigned long now=0,next_update=0;
+  now=micros();
+  if (next_update < now+now+REFRESH_INTERVAL) { // time to poll our buttons again!
+	next_update=now+REFRESH_INTERVAL;  // no matter how long it takes us to do all the crap here, our next poll will be ASAP after the refresh time expires.
```
What is "now+now+REFRESH_INTERVAL"? Why do we need to add it twice?

Also not sure how this is supposed to work. Looks like next_update is always going to be less than now+now+REFRESH_INTERVAL?

babyeatingpsychopath replied:
typo.  i hit "paste" in the wrong spot.  It should be
if (next_update < now) {
next_update=now+REFRESH_INTERVAL;

i'm not compiling this, so i have nothing to go by on whether or not it actually works.

CoryParsnipson replied:
Ah ok, I see what you're doing now.

Do you know what the trade-off here is between running a tight loop and going to low power mode? I'm thinking of doing something similar but instead we take now=micros() and then elapsed=micros() at the end of the loop, and then delay(REFRESH_INTERVAL - (elapsed - now)) so the loop always runs REFRESH_INTERVAL microseconds apart. I think we can even increase REFRESH_INTERVAL from 1ms to something like 50ms to save more power since the time between human-speed inputs is O(100ms)-ish.

This is assuming delay() puts it in low power state until it wakes up. I'd prefer to save battery life doing that especially if we're only doing a single threaded task.

CoryParsnipson replied:
Saw your post in the project and arduino threads. Gonna leave it with the idle spin loop for now and then come back and add in sleep and cut the led and other Arduino low power mods if I feel like it.

CoryParsnipson commented on code/controller/controller.ino:
Does this need to be deleted because the next_update code is added?

babyeatingpsychopath replied:
correct.  the loop function looks like this now:
loop () {
if (time_to_do_a_thing) { do things }
}
loop() will, by definition, just keep looping, letting stuff happen in the background.  no need to delay, because it just won't DO ANYTHING until it's time.

CoryParsnipson replied:
Done

CoryParsnipson commented on code/controller/controller.ino:
```diff
@@ -148,31 +141,27 @@ void reset() {
-unsigned long readAnalog(int xPin, int yPin, const int XMIN, const int XMAX, const int YMIN, const int YMAX, bool returnRaw = false) {
-  int readX = analogRead(xPin); // store in temporary variables to use in constrain()
-  int readY = analogRead(yPin); 
+int16_t readAnalog(int pin, const int MIN, const int MAX, bool returnRaw=false) {
```
This is much better than trying to return both values packed together. I like this better than the old one. Small nitpick: in this case, I'd prefer for it to be called "readAxis". Also no need to capitalize MIN and MAX.

babyeatingpsychopath replied:
yeah.  i kept as much of your old stuff as possible.  i was also confused as to why you were packing and then immediately unpacking.

CoryParsnipson replied:
Haha, I was refactoring it late at night and I wanted a single function call to return both x and y axes. Wasn't worth it.

CoryParsnipson commented on code/controller/controller.ino:
```diff
@@ -148,31 +141,27 @@ void reset() {
-    // clamp values to observed joystick values
-    readX = constrain(readX, XMIN, XMAX);
-    readY = constrain(readY, YMIN, YMAX);
+  read=analogRead(pin);
```
Can we do:

if (returnRaw) {
  return read;
}

read = constrain(read, min, max);
return map(read, min, max, -32767, 32767);

babyeatingpsychopath replied:
absolutely.  again, I was keeping the program flow largely unchanged.
i originally had it as 
if(returnRaw) return map(analogRead(pin),min,max));
return map(constrain(analogRead(pin),min,max),min,max,-32767,32767));
but felt that was tooooo much butchering.

CoryParsnipson replied:
Don't worry about butchering or keeping the original programm flow. Changes that make the code simpler (or more complex but with some other benefit that makes it worth it) matter more.

CoryParsnipson reviewed (commented):
Please see other comments for changes.

I think normally it would be helpful to group these changes into small categories and open a separate pull request for each. e.g.) PR for delay refactoring, PR for analogRead refactor, PR for prototype changes

CoryParsnipson reviewed (changes requested):
Changes requested, see other comments

CoryParsnipson commented on code/controller/controller.ino:
```diff
@@ -62,80 +64,71 @@ void setup() {
+		byte col = COL_PINS[c];
+		pinMode(col, OUTPUT);
+		digitalWrite(col, LOW);
+		//delayMicroseconds(0);  // perhaps needed to wait for the pin to settle.
```
Remove this line please

babyeatingpsychopath replied:
this is just an example of where i have, in the past, needed to put a delay() to get the pins to settle.  I was dealing with some massively crappy switches with (i suspect) a lot of line capacitance.

CoryParsnipson replied:
Oh. Hmmm, weird. I'm gonna put in a schmitt trigger (yeah I haven't forgotten) on this next time around so hopefully we won't have to worry about it. So far this version doesn't seem to have debounce issues either.

CoryParsnipson commented on code/controller/controller.ino:
```diff
@@ -62,80 +64,71 @@ void setup() {
+		  byte current_button = c * NUM_COLS + r + 1;
+		  
+		  pinMode(row, INPUT_PULLUP);
+  		  //delayMicroseconds(0);  // perhaps needed to wait for the pin to settle.
```
Remove this line please

CoryParsnipson commented on code/controller/controller.ino:
```diff
@@ -62,80 +64,71 @@ void setup() {
+  		  //delayMicroseconds(0);  // perhaps needed to wait for the pin to settle.
+		  bool button_state = !digitalRead(row);
+
+		  pinMode(row, INPUT);    // disable the pullup resistor
```
I think we can remove this line and the two lines after it for simplicity. This was added because I had connected two of these nodes to the gates of some transistors, but they aren't connected anymore.

babyeatingpsychopath replied:
ok.  the arduino pinMode() help page and the digital pins explanation page have stuff in there about order-of-operations when writing pins after setting modes.  In general, if you turn the pullups on, try to turn them off before writing the pin to make sure you're putting the value on the wire that you think you are, and not trying to output a signal through a pullup resistor.

CoryParsnipson replied:
Ok, good to know. I learned that you had to worry about this during the power switch stuff. I'll try to remember in case this comes up later.

babyeatingpsychopath committed:
changes re: comments

I'll split the major differences out into different PRs next time.  This should address all the comments.

CoryParsnipson reviewed (changes requested):
One more point of clarification needed on the delay loop stuff. (Wondering if we can add delay() back in and if it's the same thing you mentioned about putting the device in low power mode)
